dbms/src/DataStreams/ParallelInputsProcessor.h

```diff
@@ -205,27 +217,52 @@ class ParallelInputsProcessor
 
         try
         {
-            while (!finish)
-            {
-                InputData unprepared_input;
-                {
-                    std::lock_guard lock(unprepared_inputs_mutex);
+            loop(thread_num);
+        }
+        catch (...)
+        {
+            exception = std::current_exception();
+        }
 
-                    if (unprepared_inputs.empty())
-                        break;
+        if (exception)
+        {
+            handler.onException(exception, thread_num);
+        }
 
-                    unprepared_input = unprepared_inputs.front();
-                    unprepared_inputs.pop();
-                }
+        handler.onFinishThread(thread_num);
 
-                unprepared_input.in->readPrefix();
+        if (additional_inputs_at_end.empty())
+        {
+            if (0 == --active_threads)
+            {
+                handler.onFinish();
+            }
+        }
 
-                {
-                    std::lock_guard lock(available_inputs_mutex);
-                    available_inputs.push(unprepared_input);
-                }
+        {
+            std::unique_lock lock(running_first_mutex);
+            if (0 == --running_first)
+            {
+                /// Only one thread can go here so don't need to hold `unprepared_inputs_mutex`
+                /// or `unprepared_inputs_mutex` lock.
+                /// If a error happens, the `unprepared_inputs` and `available_inputs` may not be empty.
+                unprepared_inputs = UnpreparedInputs{};
+                available_inputs = AvailableInputs{};
+                for (size_t i = 0; i < additional_inputs_at_end.size(); ++i)
+                    unprepared_inputs.emplace(additional_inputs_at_end[i], i);
+
+                wait_first_done.notify_all();
             }
+            else
+            {
+                wait_first_done.wait(lock, [this] {
+                    return running_first == 0;
+                });
+            }
+        }
 
+        try
```

Comment:
L264-L276 is repeating L-218-L232, how about extract them into a function or lambda?